### PR TITLE
Integrate MaterialBanner with the ScaffoldMessenger API

### PR DESCRIPTION
**Pull request by Calamity210**
This PR integrates MaterialBanner with the ScaffoldMessenger API. This allows us to manage how many MaterialBanners are shown (at most one), as well as control the animated in/out.

**A detailed description of the reason for this change is included in the design doc.**

## Related issues
#60024

## Results

https://user-images.githubusercontent.com/50059339/117470974-5c2c7400-af25-11eb-91eb-4790b4f48857.mp4

### Hero Transition

https://user-images.githubusercontent.com/50059339/117472130-921e2800-af26-11eb-9c9e-87b7fe06a944.mp4

## Tests
Existing MaterialBanner tests have migrated and are passing, added a new control test.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

## Breaking Change
 

- [x] Yes, this is a breaking change.
- - [x] I wrote a [design doc](https://flutter.dev/go/material-banner-updates)
- - [x] I wrote a [migration guide](https://github.com/Calamity210/website/blob/master/src/docs/release/breaking-changes/material-banner-scaffold-messenger-integration.md)

### CODE REVIEW

**Review by Piinks — COMMENTED** [review]
Hey @Calamity210! This looks incredible! Wow! Thank you for the contribution!

Can you open access to the design doc you linked? I do not have permission to view. It would be great if you could add it to the website with the rest of our design docs. There are instructions here: [template](https://docs.google.com/document/d/1SFRO8U2toOlAaZ38dsuEU7Wm5fn41wvBCWKiwADqfmw/edit) I can see about sharing it with other contributors for feedback.

Also, I love that you've already written a migration guide! Thank you! We usually put those on the website too, you can see an example of one here: https://github.com/flutter/website/pull/5667

I'll start looking through this, let me know when the design doc is accessible. :)

**Comment by Calamity210** [comment]
> Can you open access to the design doc you linked?

Sorry, I forgot to update it, it should be good now @Piinks

>  It would be great if you could add it to the website with the rest of our design docs.
Already done!

**Comment by Calamity210** [comment]
> We usually put those on the website too

I was waiting for a review before opening a PR for that, the template mentioned not to open one days before a PR lands

File: packages/flutter/lib/src/material/banner.dart
Comment (Piinks) [inline packages/flutter/lib/src/material/banner.dart:228]: Instead of this, if the animation is null, can we just make it a static banner? That way it will still be backwards compatible.

Post-migration, we you planning to have the animation value required? or will it always be generated for the user? I am curious about the comment here.

**Comment by Calamity210** [comment]
@Piinks You may want to take a look at [this commit](https://github.com/flutter/flutter/commit/8cda0f3c0ebae4af70eba803ee3b65875f4dd655). I had added it to refrain from breaking existing code during the "opt-in" period.

**Review by Piinks — COMMENTED** [review]
Hey @Calamity210 it looks like this did break some tests in Google. There are plenty of options though. 👍 

Is there a way we can make this still work such that a static banner will still show for those that are currently using it?

We could also add an opt-in flag, so we can migrate the customers that are broken by this, and then later remove the opt-in flag.

Another option could be that we could create an AnimatedMaterialBanner as a separate class and maybe-maybenot deprecate the old class.

Let me see if I can loop in some of our Material Design experts as well to advise. :)

**Comment by Piinks** [comment]
@rami-a and @pennzht it looks like you may be familiar with the Google MaterialBanner tests that are broken by this change. This is a big update to align the banner with MD specs and make it more functional - like a SnackBar. Do you have any thoughts on it? :)

File: packages/flutter/lib/src/material/banner.dart
Comment (Piinks) [reply packages/flutter/lib/src/material/banner.dart:228]: > @Piinks You may want to take a look at [this commit](https://github.com/flutter/flutter/commit/8cda0f3c0ebae4af70eba803ee3b65875f4dd655) while running those tests. I had added it to refrain from breaking existing code during the "opt-in" period of migrating to the changes within this PR.

Right, I see that now. I think that the fallback animation is helpful, but still breaks users because just creating a banner shows nothing, since the animation controller is not being driven by the ScaffoldMessenger - maybe set the animation value to 1 if the fallback is used? Will that work?

Comment (Piinks) [reply packages/flutter/lib/src/material/banner.dart:228]: Sure! No rush at all. :)

File: packages/flutter/lib/src/material/banner.dart
Comment (Calamity210) [inline packages/flutter/lib/src/material/banner.dart:335]: Is this early return frowned upon? @Piinks

Comment (Calamity210) [reply packages/flutter/lib/src/material/banner.dart:335]: Couldn't find anything on it in the style guide; resolving this comment with the assumption  that it isnt a big deal.

File: packages/flutter/lib/src/material/scaffold.dart
Comment (Piinks) [inline packages/flutter/lib/src/material/scaffold.dart:761]: Is this snack bar or banner here? The docs are a bit confusing.

File: packages/flutter/lib/src/material/scaffold.dart
Comment (Piinks) [inline packages/flutter/lib/src/material/scaffold.dart:523]: Is this comment right?

File: packages/flutter/lib/src/material/scaffold.dart
Comment (Piinks) [inline packages/flutter/lib/src/material/scaffold.dart:315]: Update docs here

File: packages/flutter/lib/src/material/banner.dart
Comment (Piinks) [inline packages/flutter/lib/src/material/banner.dart:333]: I'm not sure if we need an opt-in phase. We could just make this backwards compatible and leave it available to work both ways. :)

File: packages/flutter/lib/src/material/banner.dart
Comment (Piinks) [inline packages/flutter/lib/src/material/banner.dart:202]: Looks like some snackbar docs followed over here. :)

**Review by Piinks — COMMENTED** [review]
This is coming along really nicely! This review is mostly pointing out some docs that need updating. Plus, I think we can avoid an opt in and maybe just leave this as backwards compatible. :)

**Comment by Calamity210** [comment]
> This is coming along really nicely! This review is mostly pointing out some docs that need updating. Plus, I think we can avoid an opt in and maybe just leave this as backwards compatible. :)

Sure, I reused some of the SnackBar documentation, looks like I forgot to make a couple changes 😅

**Comment by Piinks** [comment]
No worries! Copy pasta gets me every time! 😄

**Comment by Calamity210** [comment]
Resolved

File: packages/flutter/lib/src/material/scaffold.dart
Comment (Calamity210) [reply packages/flutter/lib/src/material/scaffold.dart]: Aside from:
```diff
- SnackBar
+ MaterialBanner
```
yes

File: packages/flutter/lib/src/material/scaffold.dart
Comment (Calamity210) [reply packages/flutter/lib/src/material/scaffold.dart]: snackbar, corrected

File: packages/flutter/lib/src/material/banner.dart
Comment (MrBirb) [inline packages/flutter/lib/src/material/banner.dart:43]: nit: a call*

Comment (Calamity210) [reply packages/flutter/lib/src/material/banner.dart:43]: Nice catch

File: packages/flutter/lib/src/material/banner.dart
Comment (Piinks) [inline packages/flutter/lib/src/material/banner.dart:64]: Since we are maintaining backwards compatibility here and allowing for static banners, maybe we can leave this sample and add a new one that uses the new functionality.

File: packages/flutter/lib/src/material/banner.dart
Comment (Piinks) [inline packages/flutter/lib/src/material/banner.dart:183]: This should document what happens if this is null/default behavior.
```suggestion
  /// The animation driving the entrance and exit of the material banner when presented by the [ScaffoldMessenger].
  final Animation<double>? animation;
```

File: packages/flutter/lib/src/material/banner.dart
Comment (Piinks) [inline packages/flutter/lib/src/material/banner.dart:186]: Same here - document what happens if this is null/default behavior. Is this still called if it is not used with the SacffoldMessenger?

File: packages/flutter/lib/src/material/banner.dart
Comment (Piinks) [inline packages/flutter/lib/src/material/banner.dart:332]: 👍

File: packages/flutter/lib/src/material/scaffold.dart
Comment (Piinks) [inline packages/flutter/lib/src/material/scaffold.dart:309]: Should we check to see here if `_materialBanners.isNotEmpty` and `_snackBars.isNotEmpty` before updating each? `_updateScaffolds` could be called for more than one reason now.

Comment (Piinks) [reply packages/flutter/lib/src/material/scaffold.dart:309]: You may end up pushing a duplicate snackbar for example if it has already been pushed and then a banner update comes in.

Comment (Piinks) [reply packages/flutter/lib/src/material/scaffold.dart:309]: Actually, this may be better handled in the Scaffold, see comment below.

File: packages/flutter/lib/src/material/scaffold.dart
Comment (Piinks) [inline packages/flutter/lib/src/material/scaffold.dart:2603]: we should check here if 

```dart
`_messengerMaterialBanner == _scaffoldMessenger!._materialBanners.first` 
```

And skip in that case to avoid duplicates based on this commit and the current set up in ScaffoldMessenger._updateScaffolds.

Comment (Piinks) [reply packages/flutter/lib/src/material/scaffold.dart:2603]: Same for Snackbars if that is done here.

File: packages/flutter/test/material/banner_test.dart
Comment (Piinks) [inline packages/flutter/test/material/banner_test.dart:12]: For these old tests that are being changed, since we are maintaining backwards compatibility for static banners, we should keep these pre-existing tests and add new ones that are testing the functionality you have added.

File: packages/flutter/test/material/banner_theme_test.dart
Comment (Piinks) [inline packages/flutter/test/material/banner_theme_test.dart:53]: Same comment for this file regarding pre-existing tests. Keeping the old tests ensures our backwards compatibility promise. Feel free to refactor to re-use code where convenient, but we should add new tests instead of re-writing the old to reflect new behavior.

**Review by Piinks — COMMENTED** [review]
This is a stellar PR! 🌟  
See comments below, this should also have a second reviewer take a look, I'll see about recruiting someone in the meantime. :)

File: packages/flutter/lib/src/material/banner.dart
Comment (Calamity210) [reply packages/flutter/lib/src/material/banner.dart:14]: Not that I am aware of, though [this](https://material.io/design/motion/speed.html#duration) was used as a reference (Material Banner would fall under a Medium transition area).

File: packages/flutter/lib/src/material/scaffold.dart
Comment (Calamity210) [reply packages/flutter/lib/src/material/scaffold.dart:2603]: Right, nice catch! Thoughts on the following snippet?
```dart
  void _updateMaterialBanner() {
    final ScaffoldFeatureController<MaterialBanner, MaterialBannerClosedReason>? messengerMaterialBanner =
    _scaffoldMessenger!._materialBanners.isNotEmpty
        ? _scaffoldMessenger!._materialBanners.first
        : null;

    if (_messengerMaterialBanner == messengerMaterialBanner) {
      setState(() {
        _messengerMaterialBanner = messengerMaterialBanner;
      });
    }
  }
```

File: packages/flutter/lib/src/material/banner.dart
Comment (Calamity210) [reply packages/flutter/lib/src/material/banner.dart:186]: Nope, it's only called when the MaterialBanner is presented through the ScaffoldMessenger

File: packages/flutter/lib/src/material/scaffold.dart
Comment (Calamity210) [reply packages/flutter/lib/src/material/scaffold.dart:2603]: > This is a stellar PR! 
> See comments below, this should also have a second reviewer take a look, I'll see about recruiting someone in the meantime. :)

Thanks! @Piinks

**Comment by Calamity210** [comment]
> This is a stellar PR! star2
See comments below, this should also have a second reviewer take a look, I'll see about recruiting someone in the meantime. :)

Thanks! @Piinks

**Comment by skia-gold** [comment]
Gold has detected about 2 new digest(s) on patchset 26.
View them at https://flutter-gold.skia.org/cl/github/81706

File: packages/flutter/lib/src/material/scaffold.dart
Comment (Piinks) [reply packages/flutter/lib/src/material/scaffold.dart]: Awesome! This looks good. Is there a test validating this? There should be a test that verifies we don't end up duplicating a currently showing snackbar when we push a banner, and vice versa in order to validate this logic and make sure we don't regress later. I think once we have that, this is ready to land! 🎉

Comment (Calamity210) [reply packages/flutter/lib/src/material/scaffold.dart]: @Piinks Had that in the back of my mind, just hadn't gotten around to it. I've added them in [the latest commit](https://github.com/flutter/flutter/pull/81706/commits/68d5eb8989714bfcd51c12d391d85531d58a9ce6).

**Review by Piinks — APPROVED** [review]
LGTM! Thank you for a wonderful contribution!

**Pull request merged**
